statsd/statsd.go

```diff
@@ -120,36 +141,103 @@ func (c *Client) watch() {
 }
 
 func (c *Client) append(cmd string) error {
-	c.Lock()
 	c.commands = append(c.commands, cmd)
 	// if we should flush, lets do it
 	if len(c.commands) == c.bufferLength {
 		if err := c.flush(); err != nil {
-			c.Unlock()
 			return err
 		}
 	}
-	c.Unlock()
 	return nil
 }
 
+func (c *Client) joinMaxSize(cmds []string, sep string, maxSize int) ([][]byte, []int) {
+	c.buffer.Reset() //clear buffer
+
+	var frames [][]byte
+	var ncmds []int
+	sepBytes := []byte(sep)
+
+	elem := 0
+	for _, cmd := range cmds {
+		needed := len(cmd)
+
+		//if a single command is too big, flush it and hope the UDP fragmented packet
+		//doesn't get dropped.
+		if needed > maxSize {
+			frames = append(frames, []byte(cmd))
+			ncmds = append(ncmds, 1)
+			continue
+		}
+
+		if elem != 0 {
+			needed = needed + len(sep)
+		}
+
+		if c.buffer.Len()+needed <= maxSize {
+			if elem != 0 {
+				c.buffer.Write(sepBytes)
+			}
+			c.buffer.WriteString(cmd)
+			elem++
+		} else {
+			tmpBuf := make([]byte, c.buffer.Len())
+			copy(tmpBuf, c.buffer.Bytes())
+			frames = append(frames, tmpBuf)
+			ncmds = append(ncmds, elem)
+			c.buffer.Reset()
+			c.buffer.WriteString(cmd)
+			elem = 1
+		}
+	}
+
+	//add whatever is left! if there's actually something
+	if c.buffer.Len() > 0 {
+		tmpBuf := make([]byte, c.buffer.Len())
+		copy(tmpBuf, c.buffer.Bytes())
+		frames = append(frames, tmpBuf)
+		ncmds = append(ncmds, elem)
+	}
+
+	return frames, ncmds
+}
+
 // flush the commands in the buffer.  Lock must be held by caller.
 func (c *Client) flush() error {
-	data := strings.Join(c.commands, "\n")
-	_, err := c.conn.Write([]byte(data))
+	frames, flushable := c.joinMaxSize(c.commands, "\n", OptimalPayloadSize)
+	var err error
+	cmdsFlushed := 0
+	for i, data := range frames {
+		_, e := c.conn.Write(data)
+		if e != nil {
+			err = e
+			break
+		}
+		cmdsFlushed += flushable[i]
+	}
+
 	// clear the slice with a slice op, doesn't realloc
-	c.commands = c.commands[:0]
+	if cmdsFlushed == len(c.commands) {
+		c.commands = c.commands[:0]
+	} else {
+		//this case would cause an eventual realloc in the future...
+		c.commands = c.commands[cmdsFlushed:]
+	}
 	return err
 }
 
 func (c *Client) sendMsg(msg string) error {
 	// if this client is buffered, then we'll just append this
+	c.Lock()
+	defer c.Unlock()
 	if c.bufferLength > 0 {
+		// return an error if message is bigger than OptimalPayloadSize
+		if len(msg) > MaxUDPPayloadSize {
+			return errors.New("message size exceeds MaxUDPPayloadSize")
```

Comment:
this is probably the right thing to do... We should include the size in the error message though - both the msg size and `MaxUDPPayloadSize`.